app/controllers/preferences_controller.rb

```diff
@@ -0,0 +1,32 @@
+class PreferencesController < ApplicationController
+  layout "site"
+
+  before_action :authorize_web
+  before_action :set_locale
+
+  authorize_resource :class => false
+
+  before_action :check_database_readable
+  before_action :check_database_writable, :only => [:update]
+
+  def show; end
+
+  def edit; end
+
+  def update
+    current_user.languages = params[:user][:languages].split(",")
```

Comment:
I think this needs a `set_locale` call to ensure that any changes are active? You removed it from the user settings but don't seem to have added it back here...

Reply:
There's a small discussion of that in the commit message for 9b8f2bb, but the main reason for removing it is that we no longer render a page if the update is successful. We redirect instead, which means on the subsequent page load, the user object is freshly loaded and the page will show in the new language. If the save is unsuccessful, the form is rendered again but I think that's appropriate to show the original language.

The one tiny edge case is that the flash message shown will be in the original language, so I could change that if you think the complexity is worth it?

Reply:
If that only needs the `set_locale` then I guess it might be reasonable but if it needs all the other nonsense you removed in that commit then probably not... I gave up even trying to understand the subtly of all that reset stuff.

Reply:
I just pushed a fix for this using `flash[:notice] = t(".success", :locale => Locale.available.preferred(Locale.list(current_user.languages)))` but I've just realised that a much better approach would be to use the "flash partials" approach from #3232 so give me a moment to refactor.

Reply:
OK, I've refactored (and force-pushed) so that we use a partial to ensure the successful flash message is shown in the updated language (simply by deferring the lookup of the flash message to the subsequent page load). I've added tests, but the important one is disabled until there's actually another translation to test against!